### PR TITLE
solution

### PR DESCRIPTION
**Pull request by tiserett**
[DEMO LINK](https://tiserett.github.io/react_carousel/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
-import Carousel from './components/Carousel';
+import { Carousel } from './components/Carousel';
 
 interface State {
-  images: string[];
+  images?: string[];
+  itemWidth?: number;
+  frameSize?: number;
+  step?: number;
+  animationDuration?: number;
```
Comment (sergii-nosachenko) [inline src/App.tsx:10]: These properties are not optional. On line 14 you declare all of them as default state values for your component.

File: src/components/Carousel.scss [hunk 2]
```diff
@@ -1 +1,37 @@
...
+  }
+}
+
+.btn-prev:hover {
+  transform: scale(1.2);
+}
+
+.btn-next:hover {
+  transform: scale(1.2);
+}
```
Comment (sergii-nosachenko) [inline src/components/Carousel.scss:37]:
```suggestion
  &__btn {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 10px 30px;
    border-radius: 20px;

    transition: transform 0.3s linear;
    
    &--prev,
    &--next {
      transform: scale(1.2);
    }
  }
}
```

Your BEM is bad, Maks, I'm judging you 😈
Use modifiers on your buttons.

File: src/components/Carousel.tsx [hunk 3]
```diff
@@ -1,18 +1,130 @@
...
+  canMovePrev: boolean;
+  canMoveNext: boolean;
+};
+
+export class Carousel extends Component<Props, State> {
+  state = {
+    translate: 0,
+    gap: 30,
+    canMovePrev: false,
+    canMoveNext: true,
```
Comment (sergii-nosachenko) [inline src/components/Carousel.tsx:24]: These default values don't prevent you against mistakes.
If you have only 3 items in total in 3 pieces frame you'll have next button active.

File: src/components/Carousel.tsx [hunk 4]
```diff
@@ -1,18 +1,130 @@
...
+        translate: -limit,
+        canMovePrev: true,
+        canMoveNext: false,
+      });
+    } else {
+      this.setState({
+        translate: nextStep,
+        canMovePrev: true,
+      });
+    }
```
Comment (sergii-nosachenko) [inline src/components/Carousel.tsx:67]: I think you combine it in one setState call.

File: src/components/Carousel.tsx [hunk 5]
```diff
@@ -1,18 +1,130 @@
...
+      canMoveNext,
+    } = this.state;
+
+    const {
+      images,
+      itemWidth,
+      frameSize,
+      animationDuration,
+    } = this.props;
+
```
Comment (sergii-nosachenko) [inline src/components/Carousel.tsx:84]: Something wrong is going on inside your handlers. Look [here](https://i.imgur.com/hLZns3S.png) and [here](https://i.imgur.com/xuM6eUb.png). Try to play with options and window width to catch these bugs.

File: src/components/Carousel.tsx [hunk 6]
```diff
@@ -1,18 +1,130 @@
...
+    return (
+      <div className="Carousel">
+        <button
+          className="Carousel__btn btn-prev"
+          type="button"
+          onClick={this.handleClickPrev}
+          disabled={!canMovePrev}
+        >
+          Prev
```
Comment (sergii-nosachenko) [inline src/components/Carousel.tsx:93]: It seems that you need to replace texts with arrows due the task condition.

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]
Maks, I've found some mistakes and bugs 🐛🐞🐜
Sorry, but you need to fix it.

File: src/components/Carousel.tsx [hunk 7]
```diff
@@ -1,18 +1,121 @@
...
+
+    const nextStep = translate - step * itemWidth - gap * step;
+
+    this.setState({
+      translate: (nextStep < -limit) ? -limit : nextStep,
+      canMovePrev: true,
+      canMoveNext: !(nextStep < -limit),
+    });
+  };
+
```
Comment (sergii-nosachenko) [inline src/components/Carousel.tsx:60]: Maks, it seems I've found the reason of bugs. Inside your handlers you determine the values from current state and props and then set new values. But! A huge issue here is that React optimizes the process of rendering, so now you don't know what values will be calculated (state can change suddenly during your code execution).
I propose you to use setState with callback immediately after you get inside your handler:
```
handler = () => this.setState((state, props) => {
  // calculations
  
  return {
    // new state object
  }
})
```

Why this is important? You can read it [here](https://uk.reactjs.org/docs/react-component.html#setstate)
Briefly, you need better control over what's going on when you want to set state.

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]
Code become better, but I found the source of your bugs.

**Review by sergii-nosachenko — APPROVED** [review]
Well done 🔥

File: src/App.tsx [hunk 8]
```diff
@@ -1,39 +1,120 @@
...
+      frameSize,
+      step,
+      animationDuration,
+    } = this.state;
+
+    const { images } = this.props;
 
     return (
       <div className="App">
         {/* eslint-disable-next-line */}
```
Comment (danylo-onishchuk) [inline src/App.tsx:60]: ?

**Review by danylo-onishchuk — APPROVED** [review]
good job at all :)

**Pull request closed without merging**